### PR TITLE
RGW: WIP for cloud restore and tier pre request checks

### PR DESCRIPTION
**Pull request by shreyanshjain7174**
Cloud Restore cli and its corresponding response for user.

* For first and repititive request 202 Accepted will be corresponding response code.
* For CloudRestored status 200 OK will be corresponding response code.
* For conflicting requests 409 Conflict corresponding response code.

TODO:
**if we re-run restore-object CLI with Days set, we should update expiry_date for those restored objects**

Signed-off-by: shreyanshjain7174 <ssanchet@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_rest.cc [hunk 1]
```diff
@@ -1458,7 +1458,6 @@ int RGWPostObj_ObjStore::get_params(optional_yield y)
   return 0;
 }
 
-
```
Comment (soumyakoduri) [inline src/rgw/rgw_rest.cc]: you may reset this file.

File: src/rgw/rgw_rest_s3.cc [hunk 2]
```diff
@@ -3452,6 +3452,106 @@ int RGWPostObj_ObjStore_S3::get_encrypt_filter(
...
+  if (attr_iter == attrs.end() || restore_status != rgw::sal::RGWRestoreStatus::None) {
+    s->err.http_ret = 202; //Accepted
+    dump_header(s, "x-amz-restore", rgw_bl_str(restore_status));
+  }
+  else if (attr_iter == attrs.end() || restore_status != rgw::sal::RGWRestoreStatus::RestoreAlreadyInProgress) {
+    s->err.http_ret = 409; // Conflict
+    dump_header_if_nonempty(s, "x-amz-restore", rgw_bl_str(restore_status));
+  } else if (attr_iter == attrs.end() || restore_status != rgw::sal::RGWRestoreStatus::CloudRestored) {
+    s->err.http_ret = 200; // OK
+    dump_header_if_nonempty(s, "x-amz-restore", rgw_bl_str(restore_status));
```
Comment (soumyakoduri) [inline src/rgw/rgw_rest_s3.cc]: These checks do not look correct. Why is the check "attr_iter == attrs.end()" repeated in every if case.. Its already checked once in the first if{}

else If (restore_status == rgw::sal::RGWRestoreStatus::RestoreAlreadyInProgress) { http_ret = 409 } 
else if (restore_status == rgw::sal::RGWRestoreStatus::CloudRestored) { http_ret = 200 }

Comment (shreyanshjain7174) [reply src/rgw/rgw_rest_s3.cc]: right, my bad I had took this change from my draft PR when I was testing, need to revert it to simple restore_status check.

File: src/rgw/rgw_rest.cc (hunk 1 above)
Comment (shreyanshjain7174) [reply src/rgw/rgw_rest.cc]: addressed

File: src/rgw/rgw_rest_s3.cc (hunk 2 above)
Comment (shreyanshjain7174) [reply src/rgw/rgw_rest_s3.cc]: Done!

File: src/rgw/rgw_rest_s3.cc [hunk 3]
```diff
@@ -3452,6 +3452,105 @@ int RGWPostObj_ObjStore_S3::get_encrypt_filter(
...
+
+  try {
+    RGWXMLDecoder::decode_xml("RestoreRequest", request, &parser);
+  }
+  catch (RGWXMLDecoder::err &err) {
+    ldpp_dout(this, 5) << "Malformed restore request: " << err << dendl;
+    return -EINVAL;
+  }
+
+  std::optional<uint64_t> expiry_days;
```
Comment (soumyakoduri) [inline src/rgw/rgw_rest_s3.cc]: expiry_days is defined local to this routine.. it should be defined under the class "RGWRestoreObj". get_params() can then update this common variable. And later in Execute(), you cna read expiry_days value.

**Pull request merged**